### PR TITLE
[DOC] Add Doxygen documentation support

### PR DESCRIPTION
**Pull request by xgupta**
This PR is for bootstrapping the Doxygen config by following the LLDB and MLIR examples. Not sure if it is needed today as there seems no CIRCT website to host the Doxygen docs. But I am using it personally, If the other devs found it useful please review it.

After this PR merge, we will be able to generate Doxygen doc by adding -CIRCT_INCLUDE_DOCS=ON flag to CMake invocation and then building doxygen-circt target.

### CODE REVIEW

**Comment by lattner** [comment]
Awesome.  We do have circt.org that is currently just redirecting to github, but we could use that to host something.  I think MLIR is using github pages.

I'm not a very cmake competent person.  @stephenneuendorffer could you review this?

-Chris

File: CMakeLists.txt [hunk 1]
```diff
@@ -186,11 +186,11 @@ add_subdirectory(tools)
 add_subdirectory(test)
 add_subdirectory(integration_test)
 
-#option(CIRCT_INCLUDE_DOCS "Generate build targets for the CIRCT docs."
-#  ${LLVM_INCLUDE_DOCS} ${MLIR_INCLUDE_DOCS})
-#if (CIRCT_INCLUDE_DOCS)
-#  add_subdirectory(docs)
-#endif()
+option(CIRCT_INCLUDE_DOCS "Generate build targets for the CIRCT docs."
+  ${LLVM_INCLUDE_DOCS} ${MLIR_INCLUDE_DOCS})
```
Comment (stephenneuendorffer) [inline CMakeLists.txt:190]: I'm not sure what the expected behavior here is.  What happens if LLVM_INCLUDE_DOCS=on and MLIR_INCLUDE_DOCS=on?

**Review by stephenneuendorffer — APPROVED** [review]
Thanks for this!

File: CMakeLists.txt (hunk 1 above)
Comment (xgupta) [reply CMakeLists.txt:190]: I just uncomment these lines, Currently, CIRCT is out of llvm-project repository and it does not depend on the LLVM_INCLUDE_DOCS and MLIR_INCLUDE_DOCS to be set ON. So we can remove these variables for now.

**Comment by xgupta** [comment]
Thanks for reviewing!

**Comment by xgupta** [comment]
> We do have circt.org that is currently just redirecting to github, but we could use that to host something. I think MLIR is using github pages.
>
Yes, github pages with some beautiful [hugo themes](https://themes.gohugo.io/) is nice combination. 
Where should be the right place to have that website? There is one https://github.com/circt organization on Github.

**Comment by lattner** [comment]
I'm not really sure how best to set this up. @joker-eph do you know who is the defacto owner of the MLIR web page?  Should CIRCT related stuff be located in a subsection of mlir.llvm.org or should it be its own thing?

**Comment by joker-eph** [comment]
As you prefer, I can assist to setup a circt.llvm.org on the model of mlir.llvm.org if you'd like.

**Comment by stephenneuendorffer** [comment]
> As you prefer, I can assist to setup a circt.llvm.org on the model of mlir.llvm.org if you'd like.

I think that would be great!

**Comment by lattner** [comment]
I'd love the help, thanks @joker-eph !

**Comment by joker-eph** [comment]
https://circt.llvm.org is up and running. Doxygen and other generated docs isn't plumbed it yet.
If someone wants to give it a try, just adjust this commented out section in the CIRCT deploy script, it is how it works for the MLIR website: https://github.com/llvm/circt-www/blob/main/.github/workflows/main.yml#L23-L45

**Comment by xgupta** [comment]
Thank you, @joker-eph for setting it up. Now Doxygen and other generated docs from the source code are also available on https://circt.llvm.org/

**Comment by lattner** [comment]
Awesome, it looks like we still have some work to do to spiff the output up a bit.  For example, we should have top level dialect names that show up in more of the [dialect descriptions](https://circt.llvm.org/docs/Dialects/firrtl/).  The TOC on the left isn't working fully yet (this is a content problem, not an infra problem).

**Pull request merged**